src/transport/tcp/mod.rs

```diff
@@ -129,6 +131,34 @@ pub(crate) struct TcpTransport {
     /// Connections which have been opened and negotiated but are being validated by the
     /// `TransportManager`.
     pending_open: HashMap<ConnectionId, NegotiatedConnection>,
+
+    /// Tcp metrics.
+    metrics: Option<TcpMetrics>,
+}
+
+/// TCP specific metrics.
+struct TcpMetrics {
+    /// The following metrics are used for the transport itself.
+    pending_dials_num: MetricGauge,
+    pending_inbound_connections_num: MetricGauge,
+    pending_connections_num: MetricGauge,
+    pending_raw_connections_num: MetricGauge,
+    open_raw_connections_num: MetricGauge,
+    cancel_futures_num: MetricGauge,
+    pending_open_num: MetricGauge,
+
+    /// The following metrics are shared with all TCP connections.
+    active_connections_num: MetricGauge,
+    pending_substreams_num: MetricGauge,
+}
```

Comment:
As with other metrics, I would may be reconsider what we need and what we don't. For example, I am not sure wee need the sizes of all internal containers. At the same time, the total number of substreams may be of interest.